### DISCUSSION TITLE
shrinkwrapping of drivers

### DISCUSSION THREAD

**Discussion by boneskull** [appium/appium#17112] [Ideas]
Currently, we seem to be shrinkwrapping some nonzero amount of official drivers.  For reasons, installing a driver using `appium driver install` with a recent version of npm installed (v7+) causes both the driver and `appium` itself to be installed into `$APPIUM_HOME/node_modules`.  

This could be:

- a bug in `npm install`
- the intended behavior in `npm install`
- user error when calling `npm install`
- some sort of user error when shrinkwrapping
- use of "incompatible" npm version when shrinkwrapping

I'm not sure which.  In my testing, salient flags to `npm install` have no effect on its behavior--`appium` will always be installed ostensibly because it's in the shrinkwrap file.  

But we should diagnose this and implement a workaround--if we think this is a bug in `npm`, we could report it, but we cannot rely on them to fix it in a timely manner, even if they agree it's a bug. A workaround may include, e.g.:

- further preparation of the module before shrinkwrapping such as deleting a `node_modules/appium`
- using different flags for `npm shrinkwrap`
- post-processing of the shrinkwrap file
- avoiding shrinkwrap altogether

**Comment by boneskull** [comment]
cc @mykola-mokhnach @KazuCocoa

**Comment by boneskull** [comment]
Also, another thing we could do is change installation of extensions from `$APPIUM_HOME/node_modules` into whatever directory `appium` lives in (typically a global install). This would be the same as `npm install -g appium appium-foo-driver`.  I'd want to retain `$APPIUM_HOME` as a place to keep configuration (or put this in a system-appropriate place).

**Comment by mykola-mokhnach** [comment]
> Currently, we seem to be shrinkwrapping some nonzero amount of official drivers. For reasons, installing a driver using appium driver install with a recent version of npm installed (v7+) causes both the driver and appium itself to be installed into $APPIUM_HOME/node_modules.

Which driver did you try? I have recently excluded all explicit appium dependencies from all drivers and after https://github.com/appium/appium/pull/17083 is merged I believe it should not be an issue anymore. We still import some modules in devDependencies, `@appium/test-support` or use  `@appium/support` in dependencies, but I believe this should not fetch the whole appium package.

**Reply by boneskull** [reply]
Did this just now:

![image](https://user-images.githubusercontent.com/924465/176549680-37e849af-ee55-47ae-9b49-8307866320cd.png)

**Reply by boneskull** [reply]
`appium` is explicitly in xcuitest's `npm-shrinkwrap.json` which causes it to be installed.

**Reply by boneskull** [reply]
I've been unable to determine a workaround for this case (though I haven't looked at it too closely). if using an older version of npm, `appium` doesn't get installed.  in _that_ case, we can use a symlink in `$APPIUM_HOME/node_modules/` to point to the (presumably) globally-installed `appium` (see #17083)

**Reply by boneskull** [reply]
_if_ `appium-xcuitest-driver` were installed _alongside `appium`, in the global directory_, then it would satisfy the peer dependency and (hopefully) would not get installed again.

**Reply by boneskull** [reply]
I think I'm missing the intent of keeping extensions in a user directory (`~/.appium`) vs. global installation.  A use case for this would be something like: a user wants to run a specific set of extensions which differ from the globally-installed ones in a multiuser system.  I'm not sure how Appium "typically" gets deployed, so don't know if this is a common configuration.  @jlipps ?

**Reply by mykola-mokhnach** [reply]
@KazuCocoa Could that be an issue with the shrinkwrapping release script that it tries to also shrinkwrap peer dependencies while it should not?

**Reply by KazuCocoa** [reply]
Hm... possibly? We can remove shrinkwrap if package-lock.json can be an alternative to lock install dependencies with newer npm versions, and package-lock can handle it expectedly.

Afaik, shrinkwrap was to lock dependencies in drivers' installation. I don't have any preferences to keep shrinkwrap if another better way exists.

**Reply by jlipps** [reply]
> I think I'm missing the intent of keeping extensions in a user directory (~/.appium) vs. global installation

Yeah, one design goal was to have multiple Appium configurations in different directories so sysadmins could start Appium with different configs living side by side simply by changing Appium home.

**Reply by jlipps** [reply]
> We can remove shrinkwrap if package-lock.json can be an alternative to lock install dependencies

This could be the way to go if shrinkwrapping has a bug. Or yeah, `rm -rf node_modules/appium` before shrinkwrapping.

**Reply by KazuCocoa** [reply]
oh, i see. I can push a version with the `rm -rf node_modules/appium` to check if it will work later

**Reply by KazuCocoa** [reply]
What about xcuitest `4.7.1`? The version has only product dependencies without `appium` module.
If it looks good, I'll update other drivers as same.

With the version, `appium` no longer exists in `~/.appium/appium-xcuitest-driver/node_modules/appium` but not in `~/.appium/appium-xcuitest-driver/node_modules/appium-xcuitest-driver/appium`

**Reply by RSM-SBA** [reply]
@KazuCocoa  could this be related to 4.7.1 of XCUI driver? had this errors yesterday evening.

Error: ✖ Encountered an error when installing package: npm command 'install --save-dev appium-xcuitest-driver --json' failed with code 1.

STDOUT:
npm ERR! Cannot read properties of null (reading 'pickAlgorithm')
{
  "error": {
    "code": null,
    "summary": "Cannot read properties of null (reading 'pickAlgorithm')",
    "detail": ""
  }
}

**Reply by KazuCocoa** [reply]
Potentially...? What environment did you have?
I called the below from a clean place (I mean the place did not have the driver) with NodeJS v14 and npm v8.10.0.
```
$ npm install --save-dev appium-xcuitest-driver --json
{
  "added": 920,
  "removed": 0,
  "changed": 0,
  "audited": 921,
  "funding": 64,
  "audit": {
    "vulnerabilities": {
      "info": 0,
      "low": 0,
      "moderate": 0,
      "high": 17,
      "critical": 0,
      "total": 17
    },
    "dependencies": {
      "prod": 1,
      "dev": 920,
      "optional": 206,
      "peer": 629,
      "peerOptional": 0,
      "total": 920
    }
  }
}
```

**Reply by RSM-SBA** [reply]
i do not a global installation, i install it to an appium_home directory within my project directly beside appium installation in node_modules (with node 16) . But i try to investigate a bit

**Reply by RSM-SBA** [reply]
still confused and cannot find the reason why the installation works on my machine but not on the CI, which uses the same installation procedure. Maybe the npm stack helps ?

1617 verbose stack TypeError: Cannot read properties of null (reading 'pickAlgorithm')
1617 verbose stack     at Integrity.match (/usr/local/Cellar/node@16/16.15.1/lib/node_modules/npm/node_modules/ssri/lib/index.js:273:24)
1617 verbose stack     at CachePolicy.satisfies (/usr/local/Cellar/node@16/16.15.1/lib/node_modules/npm/node_modules/make-fetch-happen/lib/cache/policy.js:112:49)
1617 verbose stack     at Function.find (/usr/local/Cellar/node@16/16.15.1/lib/node_modules/npm/node_modules/make-fetch-happen/lib/cache/entry.js:177:25)
1617 verbose stack     at async cacheFetch (/usr/local/Cellar/node@16/16.15.1/lib/node_modules/npm/node_modules/make-fetch-happen/lib/cache/index.js:8:17)
1617 verbose stack     at async fetch (/usr/local/Cellar/node@16/16.15.1/lib/node_modules/npm/node_modules/make-fetch-happen/lib/fetch.js:98:7)
1618 verbose cwd /Volumes/MyPassport/symbolLinks/HomeOfQA/workspace/RMA-iOS-Java-Releases/uitesting
1619 verbose Darwin 21.4.0
1620 verbose node v16.15.1
1621 verbose npm  v8.11.0

**Reply by RSM-SBA** [reply]
also installing it via 
npm install appium@next
appium driver install xcuitest

instead installing everything via the package.json throws this error

**Reply by KazuCocoa** [reply]
What about running `npm cache clean --force` once?

**Reply by RSM-SBA** [reply]
figured out that only one CI machine has this problem, and this machine has low diskspace on the mac mini hard drive.
There we have installed an external HDD to the mac mini and made a symbolic link to the jenkins workspace.
Maybe thats a problem now ?

npm WARN cleanup Failed to remove some directories [
npm WARN cleanup   [
npm WARN cleanup     '/Volumes/MyPassport/symbolLinks/HomeOfQA/workspace/RMA-iOS-Java-Stage-Releases/uitesting/node_modules/appium-xcuitest-driver',
npm WARN cleanup     [Error: EINVAL: invalid argument, rmdir '/Volumes/MyPassport/symbolLinks/HomeOfQA/workspace/RMA-iOS-Java-Stage-Releases/uitesting/node_modules/appium-xcuitest-driver/node_modules/lazystream'] {
npm WARN cleanup       errno: -22,
npm WARN cleanup       code: 'EINVAL',
npm WARN cleanup       syscall: 'rmdir',
npm WARN cleanup       path: '/Volumes/MyPassport/symbolLinks/HomeOfQA/workspace/RMA-iOS-Java-Stage-Releases/uitesting/node_modules/appium-xcuitest-driver/node_modules/lazystream'
npm WARN cleanup     }
npm WARN cleanup   ]
npm WARN cleanup ]
npm ERR! Cannot read properties of null (reading 'pickAlgorithm')
{
  "error": {
    "code": null,
    "summary": "Cannot read properties of null (reading 'pickAlgorithm')",
    "detail": ""
  }
}

**Reply by RSM-SBA** [reply]
with xcuitest driver 4.7.0 it works on that machine nwith symbolic linked folders

**Reply by RSM-SBA** [reply]
ok removed the symbolic link, but it still not work to install 4.7.1 on that machine... i'm out of ideas

**Reply by RSM-SBA** [reply]
[npm.log](https://github.com/appium/appium/files/9037253/npm.log)

**Reply by RSM-SBA** [reply]
ok found the problem
1) doing "npm cache clean --force"   --> no effect
2) reinstalling node 16 --> no effect
3) deleting .npmrc.  --> works now

Seems that the .npmrc was anyhow corrupt

**Reply by KazuCocoa** [reply]
Thank you.
Could you share the log wihtout `.npmrc`? Then, could you create an issue...? (As this thread is a different topic)

**Reply by RSM-SBA** [reply]
Hmm really an issue ? I don't know if appium is responsible for breaking it, because all other machines (which are all identical) worked fine. Also there are some older threads on stackoverflow which also hat asked for help regarding this error (Cannot read properties of null (reading 'pickAlgorithm')). So i think its more a npm problem.

**Comment by boneskull** [comment]
@jlipps 

> Yeah, one design goal was to have multiple Appium configurations in different directories so sysadmins could start Appium with different configs living side by side simply by changing Appium home.

What do you mean by "configurations"?

**Reply by jlipps** [reply]
different sets of drivers and plugins
